### PR TITLE
feat: add horizontal headers

### PR DESCRIPTION
**Pull request by enell**
Refs: VIZ-2530

### CODE REVIEW

**Comment by a-m-dev** [comment]
Just a quick question about the Thick borders, I believe you kept it for another PR right?

Screenshot 2023-11-03 at 10 39 35

**Comment by a-m-dev** [comment]
I'm quite sure that we discussed about this scenario, probably you kept for next one right?  

https://github.com/qlik-oss/sn-pivot-table/assets/29652890/cc9189f7-a6b6-4765-a66e-543481176423

**Comment by enell** [comment]
> Just a quick question about the Thick borders, I believe you kept it for another PR right?
> 
> 
> 
> Screenshot 2023-11-03 at 10 39 35
> 
> 

Correct. I will do styling later

**Comment by a-m-dev** [comment]
I believe we have fixed this issue in one of the prs, but it's good to keep and eye on it to not re-introduce it



https://github.com/qlik-oss/sn-pivot-table/assets/29652890/d822710a-256b-4fee-a5b0-0fc27993055d

**Comment by a-m-dev** [comment]
If i try to move all dims into columns (in order to make them all stacked on top of each other), I will get some problem while resizing and there will be free space in Headers grid


https://github.com/qlik-oss/sn-pivot-table/assets/29652890/d974fc5c-ca13-4340-98dc-11ef0e139c87

**Comment by a-m-dev** [comment]
I think (maybe) the free row scenario also causes another issue on scrolling -> it makes left grid and data grid to be out of sync when it comes to the end of page

https://github.com/qlik-oss/sn-pivot-table/assets/29652890/2ef00d83-a75e-4e6e-9a9e-8d38e6d1b6be

File: src/pivot-table/components/cells/DimensionTitleCell.tsx [hunk 1]
```diff
@@ -89,7 +89,7 @@ const DimensionTitleCell = ({
       isDimension={cell.isDim}
       style={{
         ...style,
-        ...getBorderStyle(true, isLastColumn, styleService.grid.border),
+        ...getBorderStyle(isLastRow, isLastColumn, styleService.grid.border),
```
Comment (cbt1) [inline src/pivot-table/components/cells/DimensionTitleCell.tsx:92]: `getBorderStyle` no longer includes the `cellStyle`. You will need to importat that as `baseCellStyle`

File: src/pivot-table/data/extract-headers.ts [hunk 2]
```diff
@@ -1,59 +1,111 @@
...
-        isLocked: qDimensionInfo.qLocked ?? false,
-        columnWidth: qDimensionInfo.columnWidth,
-        qApprMaxGlyphCount: qDimensionInfo.qApprMaxGlyphCount,
-        isDim: true,
-        headTextAlign: "left",
-        dimensionInfoIndex: getDimensionInfoIndex(qDimensionInfo),
-      };
-    }
+  const id = getKey(qDimensionInfo);
+  if (qDimensionInfo === PSEUDO_DIMENSION_INDEX)
```
Comment (cbt1) [inline src/pivot-table/data/extract-headers.ts:36]: Lets add curly braces for the if statement

File: src/pivot-table/data/headers-data.ts [hunk 3]
```diff
@@ -6,15 +7,19 @@ const createHeadersData = (
   visibleTopDimensionInfo: VisibleDimensionInfo[],
   visibleLeftDimensionInfo: VisibleDimensionInfo[],
 ): HeadersData => {
-  // rowCount cannot be 0, as it couse issue when there is no top data but there is left data
   const data = extractHeaders(layoutService, visibleTopDimensionInfo, visibleLeftDimensionInfo);
-
+  const size = {
+    x: data[0]?.length || 0, // number of columns
```
Comment (cbt1) [inline src/pivot-table/data/headers-data.ts:12]: Nitpicking but this should be the nullish coalescing operator
```suggestion
    x: data[0]?.length ?? 0, // number of columns
```

File: src/pivot-table/data/headers-data.ts [hunk 4]
```diff
@@ -6,15 +7,19 @@ const createHeadersData = (
...
+    y: data.length, // number of rows
+  };
+  const pseudoDimensionColumnIdx = visibleLeftDimensionInfo.reduce(
```
Comment (cbt1) [inline src/pivot-table/data/headers-data.ts:15]: Use `findIndex` instead?

File: src/pivot-table/hooks/use-column-width.ts [hunk 5]
```diff
@@ -112,34 +105,46 @@ export default function useColumnWidth(
...
+                  const fitToContentWidth = measureTextForDimensionValue(qFallbackTitle) + TOTAL_CELL_PADDING;
+                  return getColumnWidth(columnWidth, fitToContentWidth);
+                }),
+              );
+            } else if (cell.id !== PSEUDO_DIMENSION_KEY) {
+              const { qApprMaxGlyphCount, columnWidth } = cell;
+              const expandIconSize = !isFullyExpanded && colIdx < qNoOfLeftDims - 1 ? EXPAND_ICON_SIZE : 0;
+              const fitToContentWidth = Math.max(
+                width,
+                estimateWidthForDimensionValue(qApprMaxGlyphCount as number) + expandIconSize,
```
Comment (cbt1) [inline src/pivot-table/hooks/use-column-width.ts:135]: To match `main` this should have `TOTAL_CELL_PADDING` as well

File: src/pivot-table/hooks/use-column-width.ts [hunk 6]
```diff
@@ -112,34 +105,46 @@ export default function useColumnWidth(
...
+            const lockedIconSize = isLocked ? LOCK_ICON_SIZE : 0;
+            width = measureTextForHeader(label) + TOTAL_CELL_PADDING + MENU_ICON_SIZE + lockedIconSize;
+          }
+
+          if (isLastRow) {
+            if (colIdx === headersData.pseudoDimensionColumnIdx) {
+              width = Math.max(
+                width,
+                ...qMeasureInfo.map(({ qFallbackTitle, columnWidth }) => {
+                  const fitToContentWidth = measureTextForDimensionValue(qFallbackTitle) + TOTAL_CELL_PADDING;
```
Comment (cbt1) [inline src/pivot-table/hooks/use-column-width.ts:126]: I think this is bugged (on `main` as well), it should be using something like `measureTextForMeasureLabel`. Should try and fix that once this PR is merged.

Comment (enell) [reply src/pivot-table/hooks/use-column-width.ts]: Fixed

File: src/pivot-table/hooks/use-column-width.ts (hunk 5 above)
Comment (enell) [reply src/pivot-table/hooks/use-column-width.ts]: Fixed

File: src/pivot-table/data/headers-data.ts (hunk 4 above)
Comment (enell) [reply src/pivot-table/data/headers-data.ts]: Doh... Fixed

File: src/pivot-table/components/cells/DimensionTitleCell.tsx (hunk 1 above)
Comment (enell) [reply src/pivot-table/components/cells/DimensionTitleCell.tsx:90]: Done

File: src/pivot-table/data/extract-headers.ts (hunk 2 above)
Comment (enell) [reply src/pivot-table/data/extract-headers.ts]: Done

File: src/pivot-table/data/headers-data.ts (hunk 3 above)
Comment (enell) [reply src/pivot-table/data/headers-data.ts:11]: done

File: test/rendering/render.mjs-snapshots/scenario-1-chromium-linux.png
Comment (cbt1) [inline test/rendering/render.mjs-snapshots/scenario-1-chromium-linux.png:1]: Do you know why the width of the pseudo dimension column has changed?

File: test/rendering/render.mjs-snapshots/scenario-6-firefox-linux.png
Comment (cbt1) [inline test/rendering/render.mjs-snapshots/scenario-6-firefox-linux.png:1]: The increase in width of the pseudo dimension column here seems wrong. It looks like it takes the width of the menu icon into account?

File: test/rendering/render.mjs-snapshots/scenario-pagination-footer-chromium-linux.png
Comment (cbt1) [inline test/rendering/render.mjs-snapshots/scenario-pagination-footer-chromium-linux.png:1]: I would expect this scenario not change at all?

File: src/services/layout-service.ts [hunk 7]
```diff
@@ -15,9 +15,9 @@ const createLayoutService = (
     x: isSnapshot ? snapshotDataPage.qWidth : Math.min(layout.qHyperCube.qSize.qcx, MAX_COLUMN_COUNT),
     y: isSnapshot ? snapshotDataPage.qHeight : Math.min(layout.qHyperCube.qSize.qcy, MAX_ROW_COUNT),
   };
-  const hasPseudoDimOnLeft = qEffectiveInterColumnSortOrder
-    .slice(0, qNoOfLeftDims)
-    .some((index) => index === PSEUDO_DIMENSION_INDEX);
+  const isLeftDimension = (dimensionInfoIndex: number) =>
+    qEffectiveInterColumnSortOrder.slice(0, qNoOfLeftDims).some((index) => index === dimensionInfoIndex);
```
Comment (cbt1) [inline src/services/layout-service.ts:19]: Consider moving `qEffectiveInterColumnSortOrder.slice(0, qNoOfLeftDims)` out of the function. There is no reason to slice the array on every function call.

File: src/pivot-table/hooks/use-column-width.ts [hunk 8]
```diff
@@ -102,41 +102,62 @@ export default function useColumnWidth(
...
+            if (header.id === PSEUDO_DIMENSION_KEY) {
+              // Use the max width of all measures
+              cellWidth = maxMeasureCellWidth;
+            } else {
+              const { label, qApprMaxGlyphCount, columnWidth, isLocked } = header;
+              const expandIconSize = !isFullyExpanded && collIdx < qNoOfLeftDims - 1 ? EXPAND_ICON_SIZE : 0;
+              const lockedIconSize = isLocked ? LOCK_ICON_SIZE : 0;
+
+              let fitToContentWidth = TOTAL_CELL_PADDING;
+              if (layoutService.isLeftDimension(header.dimensionInfoIndex)) {
```
Comment (cbt1) [inline src/pivot-table/hooks/use-column-width.ts:140]: Should be able to resolve and assign `layoutService.isLeftDimension(header.dimensionInfoIndex)` on the actual `header` in `extract-headers`. I will probably reduce the number of times it has to be resolved.

File: src/pivot-table/hooks/use-column-width.ts [hunk 9]
```diff
@@ -102,41 +102,62 @@ export default function useColumnWidth(
...
+                fitToContentWidth += Math.max(
+                  measureTextForHeader(label) + MENU_ICON_SIZE + lockedIconSize,
+                  estimateWidthForDimensionValue(qApprMaxGlyphCount as number) + expandIconSize,
+                );
+              } else if (
+                rowIdx === headersData.size.y - 1 &&
+                collIdx === headersData.size.x - 1 &&
+                !layoutService.isLeftDimension(header.dimensionInfoIndex) &&
+                layoutService.isLeftDimension(PSEUDO_DIMENSION_INDEX)
```
Comment (cbt1) [inline src/pivot-table/hooks/use-column-width.ts:149]: Can use `layoutService.hasPseudoDimOnLeft` ?

Comment (enell) [reply src/pivot-table/hooks/use-column-width.ts]: Yes. Fixed

File: src/pivot-table/hooks/use-column-width.ts (hunk 8 above)
Comment (enell) [reply src/pivot-table/hooks/use-column-width.ts:140]: Done

File: src/services/layout-service.ts (hunk 7 above)
Comment (enell) [reply src/services/layout-service.ts:19]: Done

File: test/rendering/render.mjs-snapshots/scenario-6-firefox-linux.png
Comment (enell) [reply test/rendering/render.mjs-snapshots/scenario-6-firefox-linux.png:1]: Fixed

File: test/rendering/render.mjs-snapshots/scenario-pagination-footer-chromium-linux.png
Comment (enell) [reply test/rendering/render.mjs-snapshots/scenario-pagination-footer-chromium-linux.png:1]: Correct. Change has been reverted

File: src/pivot-table/components/cells/EmptyHeaderCell.tsx [hunk 10]
```diff
@@ -20,7 +20,7 @@ const EmptyHeaderCell = ({ rowSpan, columnSpan }: Props) => {
       }}
       data-testid={testId}
     />
-  );
+  ) : null;
```
Comment (a-m-dev) [inline src/pivot-table/components/cells/EmptyHeaderCell.tsx:23]: Maybe more Imperative way in this case would apply nicely? or am i the only one who is a bit obsessive about  conditional rendering? 😅
```jsx
if(rowSpan == 0 && colSpan == 0) return null;

return (
   <div style={{...}} />
)
```

Comment (cbt1) [reply src/pivot-table/components/cells/EmptyHeaderCell.tsx:23]: I also find that easier to read @a-m-dev

**Review by cbt1 — APPROVED** [review]
LGTM

**Comment by a-m-dev** [comment]
I noticed a small issue: 
when you resize the last column in x axis, first it tries to show label fully and avoids to set the width to the min width (30px), but then if i do it again, it sets properly
https://github.com/qlik-oss/sn-pivot-table/assets/29652890/123110f6-461d-4c77-9eaa-ad0048ee8099

**Pull request merged**